Adding validation rules into CRDs for validation

cici37 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
3. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind feature


**What this PR does / why we need it**:
This is to add validation rules to CRDs which did the same validation as written in https://github.com/kubernetes-csi/external-snapshotter/blob/master/pkg/validation-webhook/validation.go


**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #1072 

**Special notes for your reviewer**:
The feature is available since Kubernetes 1.25. Is there any version compatibility concern there? Could I remove the existing validating webhook in the same PR? Thank you!

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Adds validation rules into CRDs. Minimum required Kubernetes version is 1.25 for these validation rules.
```

cici37 commented:
/cc @msau42 
This is the changes I mentioned earlier. Would you mind help with adding ok-to-test label?
A quick question: is that ok to remove the webhook directly from this repo? The feature is general available since 1.25, is there any concern with compatibility or others? Thank you!

carlory commented:
When I run `client/hack/update-crd.sh` on the master branch, I find the issue that the content in `client/config/crds` is overwritten and inconsistent with the script. Compared to the current state, the re-generated content is missing some fields, and some fields' definitions are changed. i.e. `additionalPrinterColumns` is wiped out and `spec` is not updated. `restoreSize` type is changed from `string` to `string` or `integer`.

We should fix the issue first and then do the following steps. Modifying the generated manifest files is not a good practice. 

/hold 

/cc @xing-yang @jsafrane

xing-yang commented:
We made manual changes and documented what need to be changed in the README after running the scripts.  It's possible that some of those changes are no longer needed.  I'll take a look.

/assign

xing-yang commented:
/ok-to-test

cici37 commented:
> @cici37 Can these validation rules be generated automatically by adding some marker in the API definitions? https://github.com/kubernetes-csi/external-snapshotter/blob/master/client/apis/volumesnapshot/v1/types.go
> 
> If not, we'll have to document these as manual steps to modify the manifest after running update-crd.sh: https://github.com/kubernetes-csi/external-snapshotter/tree/master/client/hack#update-crdsh

Thanks for the review! I have added the validation through kubebuilder marker. Do we still add the same rules under CRDs or it will be auto generated through kubebuilder? Thanks

xing-yang commented:
@cici37 I updated the manifests based on your PR: https://github.com/kubernetes-csi/external-snapshotter/pull/1076

cici37 commented:
@xing-yang Thank you for updating the manifests! Is there any other changes needed in order to merge this PR?

xing-yang commented:
@cici37 I need to do some testing.  After that we can get this PR merged.  I think webhook logic can be removed after that in a followup PR.

xing-yang commented on client/apis/volumegroupsnapshot/v1alpha1/types.go:
```diff
@@ -347,11 +355,14 @@ type VolumeGroupSnapshotContentStatus struct {
 // VolumeGroupSnapshotContentSource represents the CSI source of a group snapshot.
 // Exactly one of its members must be set.
 // Members in VolumeGroupSnapshotContentSource are immutable.
+// +kubebuilder:validation:XValidation:rule="!has(oldSelf.volumeHandle) || has(self.volumeHandle)", message="volumeHandle is required once set"
```
volumeHandle->volumeHandles

cici37 replied:
Updated :)

xing-yang commented:
I thought one of the newly added rules will allow only one of persistentVolumeClaimName and volumeSnapshotContentName to be specified in VolumeSnapshot source, but it didn't happen.  I'm running K8s 1.30.

```
---
apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshot
metadata:
  name: invalid-snapshot-demo-v1
spec:
  volumeSnapshotClassName: csi-hostpath-snapclass-v1
  source: # Only one of the two fields should be set for a snapshot. Therefore this snapshot is invalid.
    persistentVolumeClaimName: pvc
    volumeSnapshotContentName: vsc
```

```
kubectl create -f invalid-snapshot-v1.yaml
volumesnapshot.snapshot.storage.k8s.io/invalid-snapshot-demo-v1 created

kubectl get volumesnapshot invalid-snapshot-demo-v1
NAME                       READYTOUSE   SOURCEPVC   SOURCESNAPSHOTCONTENT   RESTORESIZE   SNAPSHOTCLASS               SNAPSHOTCONTENT   CREATIONTIME   AGE
invalid-snapshot-demo-v1   false        pvc         vsc                                   csi-hostpath-snapclass-v1                                    4m55s
```

cici37 commented:
Thanks for the review! 
> Does this rule only makes sure that persistentVolumeClaimName and volumeSnapshotContentName are immutable once set, but it does not make sure that only one of them can be set? If so, we'll need to keep the "oneOf" rule which was added manually unless if there's a way to add this oneOf rule using CEL?
> 
> ```
>                 oneOf:
>                 - required: ["persistentVolumeClaimName"]
>                 - required: ["volumeSnapshotContentName"]
> ```

Yes the current rule only enforce the immutability of the field. I didn't see the oneOf rule in the webhook. Maybe I missed it-_-. Will update the PR to address this use case :)

xing-yang commented:
@cici37 Those Oneof are not in the webhook.  We manually added them to the manifests.  There are 3 of them.  See below:
(1) https://github.com/kubernetes-csi/external-snapshotter/blob/v7.0.2/client/config/crd/groupsnapshot.storage.k8s.io_volumegroupsnapshotcontents.yaml#L140

```
                type: object
                oneOf:
                - required: ["volumeHandles"]
                - required: ["groupSnapshotHandles"]
```

(2) https://github.com/kubernetes-csi/external-snapshotter/blob/v7.0.2/client/config/crd/snapshot.storage.k8s.io_volumesnapshotcontents.yaml#L120
```
                type: object
                oneOf:
                - required: ["snapshotHandle"]
                - required: ["volumeHandle"]
```

(3) https://github.com/kubernetes-csi/external-snapshotter/blob/v7.0.2/client/config/crd/snapshot.storage.k8s.io_volumesnapshots.yaml#L104
```
                type: object
                oneOf:
                - required: ["persistentVolumeClaimName"]
                - required: ["volumeSnapshotContentName"]
```

leonardoce commented on client/apis/volumesnapshot/v1/types.go:
```diff
@@ -99,6 +100,8 @@ type VolumeSnapshotSpec struct {
 // object should be used.
 // Exactly one of its members must be set.
 // Members in VolumeSnapshotSource are immutable.
+// +kubebuilder:validation:XValidation:rule="!has(oldSelf.persistentVolumeClaimName) || has(self.persistentVolumeClaimName)", message="persistentVolumeClaimName is required once set"
+// +kubebuilder:validation:XValidation:rule="!has(oldSelf.volumeSnapshotContentName) || has(self.volumeSnapshotContentName)", message="volumeSnapshotContentName is required once set"
```
We could add a rule to ensure we set exactly one of the members

```suggestion
// +kubebuilder:validation:XValidation:rule="!has(oldSelf.persistentVolumeClaimName) || has(self.persistentVolumeClaimName)", message="persistentVolumeClaimName is required once set"
// +kubebuilder:validation:XValidation:rule="!has(oldSelf.volumeSnapshotContentName) || has(self.volumeSnapshotContentName)", message="volumeSnapshotContentName is required once set"
// +kubebuilder:validation:XValidation:rule="(has(self.volumeSnapshotContentName) && !has(self.persistentVolumeClaimName)) || (!has(self.volumeSnapshotContentName) && has(self.persistentVolumeClaimName))", message="exactly one of volumeSnapshotContentName and persistentVolumeClaimName must be set"
```

leonardoce commented on client/apis/volumesnapshot/v1/types.go:
```diff
@@ -108,13 +111,15 @@ type VolumeSnapshotSource struct {
 	// created.
 	// This field is immutable.
 	// +optional
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Spec.Source.PersistentVolumeClaimName is immutable"
```
I believe that doing this we would have a better error message:

```suggestion
	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="persistentVolumeClaimName is immutable"
```

The original was:

```
The VolumeSnapshot "new-snapshot-test" is invalid: spec.source.persistentVolumeClaimName: Invalid value: "string": Spec.Source.PersistentVolumeClaimName is immutable
```

cici37 replied:
I was trying to match exactly with the previous message and I agree with only showing field name makes sense since we will have fieldpath returned. Updated for all places. Thanks

leonardoce commented on client/apis/volumesnapshot/v1/types.go:
```diff
@@ -108,13 +111,15 @@ type VolumeSnapshotSource struct {
 	// This field should be set if the snapshot already exists and only needs a representation in Kubernetes.
 	// This field is immutable.
 	// +optional
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Spec.Source.VolumeSnapshotContentName is immutable"
```
```suggestion
	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="volumeSnapshotContentName is immutable"
```

leonardoce commented on client/apis/volumesnapshot/v1/types.go:
```diff
@@ -340,24 +347,29 @@ type VolumeSnapshotContentSpec struct {
 	// should be dynamically taken from.
 	// This field is immutable.
 	// +optional
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Spec.Source.VolumeHandle is immutable"
```
```suggestion
	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="volumeHandle is immutable"
```

leonardoce commented on client/apis/volumesnapshot/v1/types.go:
```diff
@@ -340,24 +347,29 @@ type VolumeSnapshotContentSpec struct {
 	// was (or should be) created.
 	// This field is immutable.
 	// +optional
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Spec.Source.SnapshotHandle is immutable"
```
```suggestion
	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="snapshotHandle is immutable"
```

leonardoce commented on client/apis/volumesnapshot/v1/types.go:
```diff
@@ -340,24 +347,29 @@ type VolumeSnapshotContentSpec struct {
 // VolumeSnapshotContentSource represents the CSI source of a snapshot.
 // Exactly one of its members must be set.
 // Members in VolumeSnapshotContentSource are immutable.
+// +kubebuilder:validation:XValidation:rule="!has(oldSelf.volumeHandle) || has(self.volumeHandle)", message="volumeHandle is required once set"
+// +kubebuilder:validation:XValidation:rule="!has(oldSelf.snapshotHandle) || has(self.snapshotHandle)", message="snapshotHandle is required once set"
```
```suggestion
// +kubebuilder:validation:XValidation:rule="!has(oldSelf.volumeHandle) || has(self.volumeHandle)", message="volumeHandle is required once set"
// +kubebuilder:validation:XValidation:rule="!has(oldSelf.snapshotHandle) || has(self.snapshotHandle)", message="snapshotHandle is required once set"
// +kubebuilder:validation:XValidation:rule="(has(self.volumeHandle) && !has(self.snapshotHandle)) || (!has(self.volumeHandle) && has(self.snapshotHandle))", message="exactly one of volumeHandle and snapshotHandle must be set"
```

leonardoce commented:
I collected a set of test cases here: https://github.com/leonardoce/volumesnapshot-cel/ 
For now, I only tested `VolumeSnapshot` and `VolumeSnapshotContent`.
Should it be helpful, we can move them where they fit best, or rewrite them, or throw them away altogether.

leonardoce commented on client/apis/volumegroupsnapshot/v1alpha1/types.go:
```diff
@@ -36,6 +36,7 @@ type VolumeGroupSnapshotSpec struct {
 	// class will be used.
 	// Empty string is not allowed for this field.
 	// +optional
+	// +kubebuilder:validation:XValidation:rule="size(self) > 0",message="VolumeGroupSnapshotClassName must not be the empty string when set"
```
```suggestion
	// +kubebuilder:validation:XValidation:rule="size(self) > 0",message="volumeGroupSnapshotClassName must not be the empty string when set"
```

cici37 replied:
Updated

leonardoce commented on client/apis/volumegroupsnapshot/v1alpha1/types.go:
```diff
@@ -53,6 +56,7 @@ type VolumeGroupSnapshotSource struct {
 	// Once a VolumeGroupSnapshotContent is created and the sidecar starts to process
 	// it, the volume list will not change with retries.
 	// +optional
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Spec.Source.Selector is immutable"
```
```suggestion
	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="selector is immutable"
```

leonardoce commented on client/apis/volumegroupsnapshot/v1alpha1/types.go:
```diff
@@ -61,6 +65,7 @@ type VolumeGroupSnapshotSource struct {
 	// only needs a representation in Kubernetes.
 	// This field is immutable.
 	// +optional
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Spec.Source.VolumeGroupSnapshotContentName is immutable"
```
```suggestion
	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="volumeGroupSnapshotContentName is immutable"
```

leonardoce commented on client/apis/volumegroupsnapshot/v1alpha1/types.go:
```diff
@@ -44,6 +45,8 @@ type VolumeGroupSnapshotSpec struct {
 // object should be used.
 // Exactly one of its members must be set.
 // Members in VolumeGroupSnapshotSource are immutable.
+// +kubebuilder:validation:XValidation:rule="!has(oldSelf.selector) || has(self.selector)", message="selector is required once set"
+// +kubebuilder:validation:XValidation:rule="!has(oldSelf.volumeGroupSnapshotContentName) || has(self.volumeGroupSnapshotContentName)", message="volumeGroupSnapshotContentName is required once set"
```
```suggestion
// +kubebuilder:validation:XValidation:rule="!has(oldSelf.selector) || has(self.selector)", message="selector is required once set"
// +kubebuilder:validation:XValidation:rule="!has(oldSelf.volumeGroupSnapshotContentName) || has(self.volumeGroupSnapshotContentName)", message="volumeGroupSnapshotContentName is required once set"
// +kubebuilder:validation:XValidation:rule="(has(self.selector) && !has(self.volumeGroupSnapshotContentName)) || (!has(self.selector) && has(self.volumeGroupSnapshotContentName))", message="exactly one of selector and volumeGroupSnapshotContentName must be set"
```

leonardoce commented on client/apis/volumegroupsnapshot/v1alpha1/types.go:
```diff
@@ -265,6 +270,9 @@ type VolumeGroupSnapshotContentSpec struct {
 	// VolumeGroupSnapshot object MUST be provided for binding to happen.
 	// This field is immutable after creation.
 	// Required.
+	// +kubebuilder:validation:XValidation:rule="has(self.name) && has(self.__namespace__)",message="both Spec.VolumeGroupSnapshotRef.Name and Spec.VolumeGroupSnapshotRef.Namespace must be set"
+	// +kubebuilder:validation:XValidation:rule="self.name == oldSelf.name",message="Spec.VolumeGroupSnapshotRef.Name is immutable"
+	// +kubebuilder:validation:XValidation:rule="self.__namespace__ == oldSelf.__namespace__",message="Spec.VolumeGroupSnapshotRef.Namespace is immutable"
```
```suggestion
	// +kubebuilder:validation:XValidation:rule="has(self.name) && has(self.__namespace__)",message="both volumeGroupSnapshotRef.name and volumeGroupSnapshotRef.namespace must be set"
	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="volumeGroupSnapshotRef is immutable"
```

leonardoce commented on client/apis/volumegroupsnapshot/v1alpha1/types.go:
```diff
@@ -347,11 +355,14 @@ type VolumeGroupSnapshotContentStatus struct {
 // Exactly one of its members must be set.
 // Members in VolumeGroupSnapshotContentSource are immutable.
+// +kubebuilder:validation:XValidation:rule="!has(oldSelf.volumeHandle) || has(self.volumeHandle)", message="volumeHandle is required once set"
+// +kubebuilder:validation:XValidation:rule="!has(oldSelf.groupSnapshotHandles) || has(self.groupSnapshotHandles)", message="groupSnapshotHandles is required once set"
```
```suggestion
// +kubebuilder:validation:XValidation:rule="!has(oldSelf.groupSnapshotHandles) || has(self.groupSnapshotHandles)", message="groupSnapshotHandles is required once set"
+kubebuilder:validation:XValidation:rule="(has(self.volumeHandles) && !has(self.groupSnapshotHandles)) || (!has(self.volumeHandles) && has(self.groupSnapshotHandles))", message="exactly one of volumeHandles and groupSnapshotHandles must be set"
```

leonardoce commented on client/apis/volumegroupsnapshot/v1alpha1/types.go:
```diff
@@ -347,11 +355,14 @@ type VolumeGroupSnapshotContentStatus struct {
 	// together. It is specified for dynamic provisioning of the VolumeGroupSnapshot.
 	// This field is immutable.
 	// +optional
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Spec.Source.VolumeHandle is immutable"
```
```suggestion
	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="volumeHandle is immutable"
```

leonardoce commented on client/apis/volumegroupsnapshot/v1alpha1/types.go:
```diff
@@ -360,6 +371,7 @@ type VolumeGroupSnapshotContentSource struct {
 	// representation was (or should be) created.
 	// This field is immutable.
 	// +optional
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Spec.Source.GroupSnapshotHandles is immutable"
```
```suggestion
	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="groupSnapshotHandles is immutable"
```

cici37 commented:
@xing-yang @leonardoce Thanks for the review! I have addressed all the comments above. Please take a look when you have time. I could have a followup PR to remove the webhook after this one merged or please feel free to let me know if someone else is gonna do so :)

xing-yang commented:
/hold cancel

xing-yang commented:
/lgtm
/approve
